doc/canid.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "CANID" "1" "February 2018" "" ""
.
.SH "NAME"
\fBcanid\fR \- the caching additional network information daemon
.
.SH "SYNOPSIS"
\fBcanid\fR [\-file \fIcachefile\fR] [\-expiry \fIsec\fR] [\-concurrency \fIn\fR] [\-port \fIport\fR]
.
.SH "DESCRIPTION"
Canid provides a simple web service for caching and simplifying information retrieved from other services (see \fIBACKENDS\fR) about Internet hostnames and IP addresses\. It is designed for use when looking up many addresses and/
.
.P
On launch, Canid begins serving on the specified port\. It shuts down cleanly on SIGINT (^C on the console)\.
.
.SH "OPTIONS"
.
.IP "\(bu" 4
\fB\-file\fR \fIcachefile\fR (default: no backing store) Use the given JSON file as a backing store for the cache\. Loads the cache from this file on startup, and saves it on termination\.
.
.IP "\(bu" 4
\fB\-expiry\fR \fIsec\fR (default: 86400, 1 day) Expire cache entries after \fIsec\fR seconds\.
.
.IP "\(bu" 4
\fB\-concurrency\fR \fIn\fR (default: 16) Allow at most \fIn\fR simultaneous pending requests per backend\.
.
.IP "\(bu" 4
\fB\-port\fR \fIport\fR (default: 8043) TCP port to listen on
.
.IP "" 0
.
.SH "RESOURCES"
Canid provides three resources via HTTP:
.
.IP "\(bu" 4
\fB/\fR
.
.IP
The root resource is an HTML page providing a simple web front\-end to the service\.
.
.IP "\(bu" 4
\fB/prefix\.json?addr=\fR
.
.IP
Look up information about the prefix associated with an address, and return it as a JSON object\. This object presently contains a \fBPrefix\fR key with the routed prefix associated with the address, an \fBASN\fR key with a BGP autonomous system number associated with the address, and a \fBCountryCode\fR key for an ISO 3166 country code associated with the address\.
.
.IP "\(bu" 4
\fB/address\.json?name=\fR
.
.IP
Look up an Internet hostname via DNS, and return the IPv4 and IPv6 addresses associated with it as a JSON object\. This object contains a \fBName\fR key with the name looked up, and an \fBAddresses\fR key containing an array of IPv4 and/or IPv6 addresses as strings\. Looking up an address for a name will cause prefix information for all addresses found to be cached, as well\.
.
.IP "" 0
.
.P
All JSON resources also contain a \fBCached\fR key, the time at which the data entry was put into the cache in [RFC3339][https://datatracker\.ietf\.org/doc/RFC3339] format\.
.
.SH "BACKENDS"
The \fBprefix\.json\fR resource currently uses the Prefix Overview and Geolocation API entry points from [RIPEstat][https://stat\.ripe\.net]\.
.
.P
The \fBaddress\.json\fR resource uses DNS, as provided by the Go standard library\'s \fBnet\.LookupIP()\fR (i\.e\., the system resolver)
.
.SH "AUTHOR"
Brian Trammell \fIbrian@trammell\.ch\fR
.
.SH "LICENSE"
Copyright (c) 2016 \- 2018 Brian Trammell
.
.P
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.
.P
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\.
.
.P
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE\.


canid.go
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"strings"
	"sync"
	"time"
)

// Structure partially covering the output of RIPEstat's prefix overview and
// geolocation API calls, for decoding JSON reponses from RIPEstat.

type RipeStatResponse struct {
	Status string
	Data   struct {
		Resource         string
		Is_Less_Specific bool
		ASNs             []struct {
			ASN int
		}
		Locations []struct {
			Country string
		}
		Block struct {
			Resource string
		}
	}
}

// Prefix information

type PrefixInfo struct {
	Prefix      string
	ASN         int
	CountryCode string
	Cached      time.Time
}

// RIPEstat backend

const ripeStatPrefixURL = "https://stat.ripe.net/data/prefix-overview/data.json"
const ripeStatGeolocURL = "https://stat.ripe.net/data/geoloc/data.json"

func callRipestat(apiurl string, addr net.IP, out *PrefixInfo) error {

	// construct a query string and add it to the URL
	v := make(url.Values)
	v.Add("resource", addr.String())
	fullUrl, err := url.Parse(apiurl)
	if err != nil {
		return err
	}
	fullUrl.RawQuery = v.Encode()

	log.Printf("calling ripestat %s", fullUrl.String())

	resp, err := http.Get(fullUrl.String())
	if err != nil {
		return err
	}

	// and now we have a response, parse it
	var doc RipeStatResponse
	dec := json.NewDecoder(resp.Body)
	err = dec.Decode(&doc)
	if err != nil {
		return err
	}

	// don't even bother if the server told us to go away
	if doc.Status != "ok" {
		return errors.New("RIPEstat request failed with status " + doc.Status)
	}

	// store the prefix, if not already present
	if len(out.Prefix) == 0 {
		if doc.Data.Is_Less_Specific {
			out.Prefix = doc.Data.Resource
		} else {
			// if the resource isn't a prefix, look for the block
			out.Prefix = doc.Data.Block.Resource
		}
	}

	// get the first AS number, if present
	for _, asn := range doc.Data.ASNs {
		out.ASN = asn.ASN
		break
	}

	// get the first country code, if present
	for _, location := range doc.Data.Locations {
		out.CountryCode = location.Country
		break
	}

	return nil
}

func lookupRipestat(addr net.IP) (out PrefixInfo, err error) {
	err = callRipestat(ripeStatPrefixURL, addr, &out)
	if err == nil {
		callRipestat(ripeStatGeolocURL, addr, &out)
	}
	return
}

// Map of prefixes to information about them, stored by prefix.

type PrefixCache struct {
	Data            map[string]PrefixInfo
	lock            sync.RWMutex
	expiry          int
	backend_limiter chan struct{}
}

func (cache *PrefixCache) lookup(addr net.IP) (out PrefixInfo, err error) {
	// Determine starting prefix by guessing whether this is v6 or not
	var prefixlen, addrbits int
	if strings.Contains(addr.String(), ":") {
		prefixlen = 48
		addrbits = 128
	} else {
		prefixlen = 24
		addrbits = 32
	}

	// Iterate through prefixes looking for a match
	for i := prefixlen; i > 0; i-- {
		mask := net.CIDRMask(i, addrbits)
		net := net.IPNet{addr.Mask(mask), mask}
		prefix := net.String()

		cache.lock.RLock()
		out, ok := cache.Data[prefix]
		cache.lock.RUnlock()
		if ok {
			// check for expiry
			if int(time.Since(out.Cached).Seconds()) > cache.expiry {
				log.Printf("entry expired for prefix %s", prefix)
				cache.lock.Lock()
				delete(cache.Data, prefix)
				cache.lock.Unlock()
				break
			} else {
				log.Printf("cache hit! for prefix %s", prefix)
				return out, nil
			}
		}
	}

	// Cache miss, go ask RIPE
	cache.backend_limiter <- struct{}{}
	out, err = lookupRipestat(addr)
	_ = <-cache.backend_limiter
	if err != nil {
		return
	}

	// cache and return
	out.Cached = time.Now().UTC()
	cache.lock.Lock()
	cache.Data[out.Prefix] = out
	cache.lock.Unlock()
	log.Printf("cached prefix %s -> %v", out.Prefix, out)

	return
}

func (cache *PrefixCache) lookupServer(w http.ResponseWriter, req *http.Request) {

	ip := net.ParseIP(req.URL.Query().Get("addr"))
	if ip == nil {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	prefix_info, err := cache.lookup(ip)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		error_struct := struct{ Error string }{err.Error()}
		error_body, _ := json.Marshal(error_struct)
		w.Write(error_body)
		return
	}

	prefix_body, _ := json.Marshal(prefix_info)
	w.Write(prefix_body)
}

// Map of names to addresses

type AddressInfo struct {
	Name      string
	Addresses []net.IP
	Cached    time.Time
}

type AddressCache struct {
	Data            map[string]AddressInfo
	lock            sync.RWMutex
	prefixes        *PrefixCache
	expiry          int
	backend_limiter chan struct{}
}

func (cache *AddressCache) lookup(name string) (out AddressInfo) {
	// Cache lookup
	var ok bool
	cache.lock.RLock()
	out, ok = cache.Data[name]
	cache.lock.RUnlock()
	if ok {
		// check for expiry
		if int(time.Since(out.Cached).Seconds()) > cache.expiry {
			log.Printf("entry expired for name %s", name)
			cache.lock.Lock()
			delete(cache.Data, name)
			cache.lock.Unlock()
		} else {
			log.Printf("cache hit for name %s", name)
			return
		}
	}

	// Cache miss. Lookup.
	out.Name = name
	cache.backend_limiter <- struct{}{}
	addrs, err := net.LookupIP(name)
	_ = <-cache.backend_limiter
	if err == nil {
		// we have addresses. precache prefix information.
		out.Addresses = addrs
		// precache prefixes, ignoring results
		for _, addr := range addrs {
			_, _ = cache.prefixes.lookup(addr)
		}
	} else {
		out.Addresses = make([]net.IP, 0)
		log.Printf("error looking up %s: %s", name, err.Error())
		err = nil
	}

	// cache and return
	out.Cached = time.Now().UTC()
	cache.lock.Lock()
	cache.Data[out.Name] = out
	cache.lock.Unlock()
	log.Printf("cached name %s -> %v", out.Name, out)
	return
}

func (cache *AddressCache) lookupServer(w http.ResponseWriter, req *http.Request) {
	// TODO figure out how to duplicate less code here
	name := req.URL.Query().Get("name")
	if len(name) == 0 {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	addr_info := cache.lookup(name)
	// if err != nil {
	// 	w.WriteHeader(http.StatusInternalServerError)
	// 	error_struct := struct{ Error string }{err.Error()}
	// 	error_body, _ := json.Marshal(error_struct)
	// 	w.Write(error_body)
	// 	return
	// }

	addr_body, _ := json.Marshal(addr_info)
	w.Write(addr_body)
}

const CANID_STORAGE_VERSION = 1

type CanidStorage struct {
	Version   int
	Prefixes  *PrefixCache
	Addresses *AddressCache
}

func (storage *CanidStorage) undump(in io.Reader) error {
	dec := json.NewDecoder(in)
	return dec.Decode(storage)
}

func (storage *CanidStorage) dump(out io.Writer) error {
	enc := json.NewEncoder(out)
	return enc.Encode(*storage)
}

func newStorage(expiry int, limit int) *CanidStorage {
	storage := new(CanidStorage)
	storage.Version = CANID_STORAGE_VERSION
	storage.Prefixes = new(PrefixCache)
	storage.Prefixes.Data = make(map[string]PrefixInfo)
	storage.Prefixes.expiry = expiry
	storage.Prefixes.backend_limiter = make(chan struct{}, limit)
	storage.Addresses = new(AddressCache)
	storage.Addresses.Data = make(map[string]AddressInfo)
	storage.Addresses.prefixes = storage.Prefixes
	storage.Addresses.expiry = expiry
	storage.Addresses.backend_limiter = make(chan struct{}, limit)
	return storage
}

func main() {
	fileflag := flag.String("file", "", "backing store for caches (JSON file)")
	expiryflag := flag.Int("expiry", 86400, "expire cache entries after n sec")
	limitflag := flag.Int("concurrency", 16, "simultaneous backend request limit")

	// parse command line
	flag.Parse()

	// set up sigterm handling
	interrupt := make(chan os.Signal, 1)
	signal.Notify(interrupt, os.Interrupt)

	// allocate and link cache
	storage := newStorage(*expiryflag, *limitflag)

	// undump cache if filename given
	if len(*fileflag) > 0 {
		infile, ferr := os.Open(*fileflag)
		if ferr == nil {
			cerr := storage.undump(infile)
			infile.Close()
			if cerr != nil {
				log.Fatal(cerr)
			}
			log.Printf("loaded caches from %s", *fileflag)
		} else {
			log.Printf("unable to read cache file %s : %s", *fileflag, ferr.Error())
		}
	}

	// check for cache version mismatch
	if storage.Version != CANID_STORAGE_VERSION {
		log.Fatal("storage version mismatch for cache file %s: delete and try again", *fileflag)
	}

	go func() {
		http.HandleFunc("/prefix.json", storage.Prefixes.lookupServer)
		http.HandleFunc("/address.json", storage.Addresses.lookupServer)
		log.Fatal(http.ListenAndServe(":8081", nil))
	}()

	_ = <-interrupt
	log.Printf("terminating on interrupt")

	// dump cache if filename given
	if len(*fileflag) > 0 {
		outfile, ferr := os.Create(*fileflag)
		if ferr == nil {
			cerr := storage.dump(outfile)
			outfile.Close()
			if cerr != nil {
				log.Fatal(cerr)
			}
			log.Printf("dumped cache to %s", *fileflag)
		} else {
			log.Fatalf("unable to write backing file %s : %s", *fileflag, ferr.Error())
		}
	}
}
